### PR TITLE
rust: compile protocol buffer bindings with Prost

### PR DESCRIPTION
**Pull request by wchargin**
Summary:
This patch adds infrastructure to compile Rust bindings for our protocol
buffer types. The compilation occurs in the build graph via a genrule,
and then the generated files are reflected into the source tree. A Bazel
test verifies that they are kept in sync.

We reflect the generated files into the source tree so that IDE tooling
like `rust-analyzer` can analyze these files more correctly and quickly.
There are ways to get `rust-analyzer` to consider generated files, but
they’re primarily meant for integration with Cargo `build.rs`, not Bazel
(though [Bazel `rules_rust` support for `rust_analyzer` is on the
roadmap][rr-ra]). Such configurations also force `rust-analyzer` to
execute the genrule-equivalent, which isn’t instant, especially when the
build dependencies have not yet been built.

Also, `rules_rust` doesn’t yet seem to support using generated files in
a `rust_library`/`rust_binary`, so this is necessary for now, anyway.

We compile the bindings in the build graph, rather than just providing a
binary that runs the compilation process, primarily because this gives
better control over the execution environment. The data dependencies
here are a bit finicky, and the genrule environment is well defined. By
contrast, a binary could be run either via `bazel run` or `bazel-bin/`,
and either at the repo root or the package directory. By using a
genrule, we only have to support one configuration.

Using a genrule also wins slightly better caching behavior: a normal
build of our Rust library doesn’t actually need to pull in `prost-build`
(though running the tests does).

The generated files are not subject to `rustfmt` because of how they’re
`include!`d into the crate. This is kind of a happy accident. If ever
`rustfmt` learns to format these files, we can fend it off again by
adding `#[rustfmt::skip]` to the module.

[rr-ra]: https://github.com/bazelbuild/rules_rust/issues/378

Test Plan:
Run `//tensorboard/data/server:update_protos_test` and note that it
passes. Then, modify `tensorboard.pb.rs` and re-run it. The test should
fail and instruct you to run `//tensorboard/data/server:update_protos`
to correct the discrepancy. Do so; the test should pass again.

Also, note in the documentation server that the `proto::tensorboard`
module exists and is well populated.

wchargin-branch: rust-compile-protos

### CODE REVIEW

**Comment by wchargin** [comment]
Retargeting to master to trigger a Travis run.

**Comment by wchargin** [comment]
Yay—it builds on CI, too! :-)

Down the line I also have integration of the whole Tonic stack, both
build time and runtime (incl. a working demo server). That didn’t
require any further configuration, and the Tonic stack was what I was
the most worried about (just because all our other deps are so
straightforward), so I think that at this point we’re pretty safe.

**Comment by wchargin** [comment]
Note to reviewers: the diff as written rolls up #4310, which is why it
looks large. Feel free to either review just the “rust: compile protocol
buffer bindings with Prost” commit or to just ignore all the other
generated files.

File: tensorboard/data/server/lib.rs [hunk 1]
```diff
@@ -19,3 +19,12 @@ pub mod masked_crc;
 
 #[cfg(test)]
 mod scripted_reader;
+
+/// Protocol buffer bindings.
+#[allow(clippy::all)]
```
Comment (stephanwlee) [inline tensorboard/data/server/lib.rs:24]: :) What does this do? Is this running clippy on the proto definitions?

**Review by stephanwlee — APPROVED** [review]
I personally prefer keeping generated files in bazel-genfiles (unless there is a great impact on productivity) but, given your rationale, I see no good way around it. LGTM.

File: tensorboard/data/server/update_protos.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+# ==============================================================================
+"""Update proto source files based on generated outputs."""
+
+import os
+import sys
+
+# Paths to "tensorboard/data/server/" in (a) the Bazel runfiles tree,
+# whence we can read data dependencies, and (b) the Git repository,
+# whither we can write output files.
+_REPO_DIR = os.path.dirname(os.readlink(__file__))
```
[the remark this answers is not in the document]
Comment (wchargin) [reply tensorboard/data/server/update_protos.py]: Indeed. Done; thanks.

File: tensorboard/data/server/update_protos.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+_BAZEL_DIR = os.path.join("tensorboard", "data", "server")
+
+_RUST_LICENSE = """\
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (wchargin) [reply tensorboard/data/server/update_protos.py:27]: Heh. Well, we don’t generally bump copyright each year, so maybe your
cred-farming will be noticed. ;-)

File: tensorboard/data/server/lib.rs (hunk 1 above)
Comment (wchargin) [reply tensorboard/data/server/lib.rs:24]: The opposite: it _suppresses_ all Clippy lint warnings inside the
generated code. There are a few different [diagnostic attributes][1]:

```rust
#[allow(dead_code)]
#[warn(dead_code)]
#[deny(dead_code)]
#[forbid(dead_code)]
```

where `deny` and `forbid` are both hard errors, but `forbid` also
prevents you from `allow`ing it in any child item. So, if you see a
crate that has `#![forbid(unsafe)]` at the top, then you’re immediately
100% sure that it has no uses of `unsafe`.

Here, `clippy::all` is a… shortcut? alias? group? (don’t know the right
term) for “all the Clippy lints”, so this is “allow any Clippy lint
without erroring or warning about it”.

Context: <https://github.com/danburkert/prost/issues/379>

[1]: https://doc.rust-lang.org/reference/attributes/diagnostics.html#lint-check-attributes

**Review by wchargin — COMMENTED** [review]
> I personally prefer keeping generated files in bazel-genfiles (unless
> there is a great impact on productivity) but, given your rationale, I
> see no good way around it. LGTM.

Agreed. Genfiles are great for many reasons, but this seems best to me
right now. Happy to rethink if facts or feelings change.

**Pull request merged**